File: 0198-house-robber/0198-house-robber.java
class Solution {

    int[] cache;

    public int dp(int idx, int[] nums){

        if(idx >= nums.length) {
            return 0;
        }

        if(cache[idx] != -1){
            return cache[idx];
        }

        cache[idx] = Math.max(nums[idx] + dp(idx + 2, nums), dp(idx + 1, nums));

        return cache[idx];
    }

    public int rob(int[] nums) {
        cache = new int[nums.length];
        Arrays.fill(cache, -1);
        return Math.max(dp(0, nums), dp(1, nums));
    }
}

File: 0198-house-robber/0198-house-robber.cpp
class Solution {
public:
    vector<int> cache;

    int dp(vector<int> nums, int idx) {
        if(idx >= nums.size()) return 0;

        if(cache[idx] != -1) return cache[idx];

        return cache[idx] = max(nums[idx] + dp(nums, idx + 2), dp(nums, idx + 1));
    }

    int rob(vector<int>& nums) {
        cache = vector<int>(nums.size(), -1);
        return dp(nums, 0);
    }
};